packages/state-transition/src/cache/epochCache.ts

```diff
@@ -549,6 +572,15 @@ export class EpochCache {
     // ```
     this.epoch = computeEpochAtSlot(state.slot);
     this.syncPeriod = computeSyncPeriodAtEpoch(this.epoch);
+
+    // To populate finalized cache and prune unfinalized cache with validators that are just initialized
+    if (this.isAfterEIP6110()) {
+      const validators = state.validators;
+      for (const index of epochTransitionCache.indicesEligibleForActivationQueue) {
+        const validator = validators.getReadonly(index);
+        this.addFinalizedPubkey(validator.pubkey, index);
```

Comment:
why validators in `indicesEligibleForActivationQueue` (which is precalculated in an epoch transition) considered finalized?

Reply:
my understanding is we only call `this.addFinalizedPubkey()` per finalized checkpoint, not after any epoch transitions.

when node has a finalized checkpoint, we should be able to access finalized checkpoint state and get `state.validators.length` from there, it's the `pivotValidatorIndex`, i.e all validator indices before that is finalized, all validator indices after that are not finalized

maybe need more discussions here to clarify this cc @wemeetagain @dapplion @g11tech

Reply:
I guess my line of thought of choosing to `addFinalizedPubkey()` during epoch transition is because calculation of the validator activation queue and the dequeue happen in `processRegistryUpdates()`. 

I think we can debate how we should define a finalized validator. I drew a [diagram](https://sequencediagram.org/index.html#initialData=C4S2BsFMAIDlIO7QG4ENwgCauAewE7QAOAFqgM6QBQNRq+oAxiHQHbDQDCZrA5pAFlIARioAjXAA9ouZJEIARAK74QfaAFEiuRiWjCAXEXw7I5cgCFwOgNYAKAJTipMuYpVrem7bv1GTjGbkCpDa5GCOVKgS0rLy0Mqq6lo6eoaojKBooLisAPqQGLwgYiAYwACeBT4kALwAYgCCAEp59QCqACrtzRp5GgAKAPKcABIA3AA6rBlZOCC51akNLW1dPX2DI6POsW4JHsk1ftDkwDiQAHRoGNh4+OSXqEREkKyYkYme3qn6ALQAPg07xkADMfr5DLtXPFgZgwRC0v5TOZmpBimd8BV2kQ7mZIqxcMAYHFCHCESlIQZoI1zPJgNSQOQCkUSlA8qCCHlZiBsgt8gBHJSQYWRaIuUmaEG4cGUpHQHl8xaFEDFUrlKqhX61aD8YB5RgqfBvfVa3R2M4XBzQADU+mgOoATFRyTLEf8gdLZcdJAYqITiTCyV73b6aXSGIzmSq1ezOfhuZlefNco5piArtAzXpFSn8jGSmUwJrjk7oEzoKC1OgQAAvSCYf1Ekn7V3e35hgBqNbuBHL5GgjFy4Uw8gb0Dw0DEMDUE5IMFzOVY0CFIsgABppghUGBK33d5Pp9BR6vhfCECQQL4K5gr+difCxBVByQVMuMABbMAukNy6CSQE23dCo-QDFtYV-Y4QPDShI2gAB1HcOArUl12gBAynAKcYBPYUzznCts2hSUgL-aDjBRcg0QxYAsRxPFyDFGIgyleE3TI6lF35JZdGmHUhw-IglGJRM5iXApJDAHiSDsPUDSNE1pIte9IAcBw+OgZ87WEW1oAERoAA08gAZQ0DQFDyRpRg0RoFA0rToAAVibQMSMg35oMQg8yizKCbUcwdcA-MwfzY9tfAqQD3Ii-zQObFjSL8xzqVpWCGX7UTeUgPIbiwHACAJeK3LC4DYugbtbnywgK0JJAh1YEcx3hRVICAA) on different phases for a new validator.

If I am not mistaken, I think what you are proposing is a validator is finalized **if epoch 1 is finalized** according to the diagram. 

We can consider a validator finalized when we reach either a) End of epoch `x` b) End of epoch `y` c) End of epoch `y+5`. Personally I prefer a. Regardless, the current code is not correct. It is doing it at end of epoch 1

Reply:
@naviechan I took some more time to think about this

> If I am not mistaken, I think what you are proposing is a validator is finalized if epoch 1 is finalized according to the diagram.

no I don't mean that

- I think it's not correct to call `addFinalizedPubkey()` on a normal epoch transition condition, because any epoch transitions could happen with orphaned blocks. For example given `n % 32 = 0`, we should not care the epoch transition happened on orphaned block n

```
    block n-1 -|- block (n+1) --- block (n+2) ...
              \
               \
               block n (orphaned)
```

- It's also not correct to base on validator lifecycle (the above sequence diagram) to call `addFinalizedPubkey()`. Any validators even not in `indicesEligibleForActivationQueue` could still be finalized once they are included in a finalized state because their indices never change

the only way I can think of is to call `addFinalizedPubkey()` once we have a new finalized state: given a new finalized state `stateN`, update validator indices from the last finalized state's validator length, i.e `this. pubkey2index.size`, to `stateN.validators.length`. This assumes we don't reuse validator indices, i.e this is not implemented soon https://eips.ethereum.org/EIPS/eip-6914

Reply:
> I guess my line of thought of choosing to addFinalizedPubkey() during epoch transition is because calculation of the validator activation queue and the dequeue happen in processRegistryUpdates().

Another counter argument to my point here is that, relying on validator lifecycle to determine if a validator is finalized or not wouldn't work because `activation_eligibility_epoch` would not be set the deposit amount is less than 32 ETH and `validator.activation_eligibility_epoch <= state.finalized_checkpoint.epoch` in `is_eligible_for_activation ()` would not include it in the activation queue. But it still needs to be added to finalized cache at some point. 

In short, active validators are all finalized, but non-active validators due to insufficient deposit amount should also be finalized.

> the only way I can think of is to call addFinalizedPubkey() once we have a new finalized state: given a new finalized state stateN, update validator indices from the last finalized state's validator length, i.e this. pubkey2index.size, to stateN.validators.length. This assumes we don't reuse validator indices, i.e this is not implemented soon https://eips.ethereum.org/EIPS/eip-6914

I think this is a viable way and certainly a simple solution but with a big assumption. 

We should also explore potential solutions that's compatible with EIP-6914 since we don't know how it would go

Reply:
I would not worry about EIP-6914 too much, EIP-6110 is 99% certain to be included in some form, while EIP-6914 is much less likely.